docs/writing-functions.md

```diff
@@ -197,6 +197,70 @@ FunctionsFramework.http "error_reporter" do |request|
 end
 ```
 
+## Shared resources
+
+Generally, functions should be self-contained and stateless, and should not use
+or share any global state in the Ruby VM. This is because serverless runtimes
+may spin up or terminate instances of your app at any time, and because a
+single instance may be running multiple functions at a time in separate threads.
+
+However, it is sometimes useful to share a resource across multiple function
+invocations that run on the same Ruby instance. For example, you might establish
+a single connection to a remote database or other service, and share it across
+function invocations, to avoid incurring the overhead of re-establishing that
+connection for every function invocation.
+
+When using a shared resource, it is important to keep three things in mind:
+
+ 1. **The shared resource should be thread-safe.** This is because serverless
+    runtimes such as Google Cloud Functions may run multiple functions at a time
+    in separate threads.
+
+ 2. **Use `FunctionsFramework.on_startup` to initialize shared resources.**
+    Do not initialize a shared resource at the top level of your app. This is
+    because serverless runtimes may load your files (and thus execute any Ruby
+    code at the top level) during deployment or at other times when it is not
+    actually preparing to execute functions. If that happens, resources may fail
+    to initialize (and thus, for example, cause deployment to fail) if they
+    expect to be running in a "serving" rather than a "building" environment.
+    Instead, initialize resources in a `FunctionsFramework.on_startup` block,
+    which the Functions Framework will call just before starting a server.
+
+    For example:
+
+    ```ruby
+    require "functions_framework"
+
+    # This local variable is lexically shared among all blocks.
+    storage_client = nil
+
+    #   Do not create the storage client here. This may run during deployment
```

Comment:
Indentation is off.

I'd expect the inverse indentation, i.e. no indent for text, indentation for code samples in comments.

Reply:
Changed.